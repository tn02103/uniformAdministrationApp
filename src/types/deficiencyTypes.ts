import { Prisma } from "@prisma/client";

export const deficiencyTypeArgs = Prisma.validator<Prisma.DeficiencyTypeFindManyArgs>()({
    select: {
        id: true,
        name: true,
        dependent: true,
        relation: true,
    }
});

export type DeficiencyType = Prisma.DeficiencyTypeGetPayload<typeof deficiencyTypeArgs>;

export interface AdminDeficiencyType extends DeficiencyType {
    resolved: number;
    active: number;
    disabledUser: string;
    disabledDate: Date;
}

export interface Deficiency {
    id?: string;
    typeId: string;
    typeName: string;
    description: string;
    comment: string;
    dateCreated?: Date;
    dateUpdated?: Date;
    dateResolved?: Date;
    userCreated?: Date;
    userUpdated?: Date;
    userResolved?: Date;
}
export interface CadetDeficiency extends Deficiency {
    fk_cadet: string;
    fk_material?: string;
    fk_uniform?: string;
}
export interface UniformDeficiency extends Deficiency {
    fk_uniform: string;
}

export type CadetInspection = {
    id?: string,
    uniformComplete: boolean,
    oldCadetDeficiencies: Deficiency[],
    newCadetDeficiencies: Deficiency[],
}

export type InspectionStatus = {
    active: false,
} | {
    active: true,
    id: string,
    date: Date,
    inspectedCadets: number,
    activeCadets: number,
}

export type InspectionInformation = {
    id: string,
    date: Date,
    cadetsInspected: number,
    newDeficiencies: number,
    activeDeficiencies: number,
    resolvedDeficiencies: number
}

export type InspectionReview = InspectionInformation & {
    cadetList: InspectionReviewCadet[],
    activeDeficiencyList: InspectionReviewDeficiency[],
}

export type InspectionReviewCadet = {
    cadet: {
        id: string,
        firstname: string,
        lastname: string,
    },
    lastInspection: {
        id: string,
        date: Date,
        uniformComplete: boolean,
    }
    activeDeficiencyCount: number,
    newlyClosedDeficiencyCount: number,
    overalClosedDeficiencyCount: number,
}

export type InspectionReviewDeficiency = {
    id: string,
    cadet: {
        id: string,
        firstname: string,
        lastname: string,
    },
    description: string,
    deficiencyType: {
        id: string,
        name?: string,
    },
    comment: string,
    dateCreated?: Date,
    new: boolean,
}
