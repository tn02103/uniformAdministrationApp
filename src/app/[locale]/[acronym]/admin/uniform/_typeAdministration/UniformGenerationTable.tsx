import { ReorderableTableBody } from "@/components/reorderDnD/ReorderableTableBody";
import { changeUniformGenerationSortOrder } from "@/dal/uniform/generation/_index";
import { useUniformTypeList } from "@/dataFetcher/uniformAdmin";
import { UniformType, UniformGeneration } from "@/types/globalUniformTypes";
import { faBars } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { useState } from "react";
import { Row, Table, Dropdown } from "react-bootstrap";
import { UniformgenerationOffcanvas } from "./UniformGenerationOffcanvas";
import { TooltipActionButton } from "@/components/Buttons/TooltipIconButton";
import { useI18n } from "@/lib/locales/client";
import { toast } from "react-toastify";

type UniformGenerationTableProps = {
    uniformType: UniformType;
}

export const UniformGenerationTable = ({ uniformType }: UniformGenerationTableProps) => {
    const { mutate } = useUniformTypeList();
    const t = useI18n();
    const [selectedGenerationId, setSelectedGenerationId] = useState<string | null>(null);

    const handleChangeSortorder = async (newArray: UniformGeneration[], itemId: string) => {
        if (newArray.length !== uniformType.uniformGenerationList?.length) return;
        const newPosition = newArray.findIndex(i => i.id === itemId);
        if (newPosition === -1) return;
        await mutate(
            changeUniformGenerationSortOrder({ id: itemId, newPosition })
        ).then(() => {
            toast.success(t('common.success.changeSortorder'));
        }).catch(() => {
            toast.error(t('common.error.actions.changeSortorder'));
        });
    };

    return (
        <div data-testid="uniform-generation-table">
            <h3 className="text-center">{t('common.uniform.generation.label', { count: 2 })}</h3>
            <hr className="my-0" />
            <Row>
                <Table>
                    <thead>
                        <tr>
                            <th></th>
                            <th>{t('common.uniform.generation.label', { count: 1 })}</th>
                            <th className={uniformType.usingSizes ? "d-none d-sm-table-cell" : ""}>{t('common.uniform.generation.outdated')}</th>
                            {uniformType.usingSizes && (
                                <th>{t('common.uniform.sizelist.label')}</th>
                            )}
                            <th><TooltipActionButton variantKey="create" onClick={() => setSelectedGenerationId('new')} /></th>
                        </tr>
                    </thead>
                    <ReorderableTableBody<UniformGeneration> items={uniformType.uniformGenerationList ?? []} itemType="UNIFORM_GENERATION" onDragEnd={handleChangeSortorder}>
                        {({ draggableRef, previewRef, isDragging, item }) => {
                            const invalid = (uniformType.usingSizes && !item.fk_sizelist)
                            return (
                                <tr
                                    key={item.id}
                                    ref={previewRef}
                                    style={isDragging ? { opacity: 0 } : undefined}
                                    aria-label={item.name}
                                >
                                    <td className={invalid ? "text-danger" : ""}>
                                        <span ref={draggableRef} aria-label="drag item" className="p-2">
                                            <FontAwesomeIcon
                                                icon={faBars}
                                                className="text-seccondary"
                                            />
                                        </span>
                                    </td>
                                    <td className={invalid ? "text-danger" : ""}>{item.name}</td>
                                    <td className={`${invalid ? "text-danger" : ""} ${uniformType.usingSizes ? "d-none d-sm-table-cell" : ""}`}>
                                        {item.outdated ? t('common.yes') : t('common.no')}
                                    </td>
                                    {uniformType.usingSizes && (
                                        <td>{item.sizelist?.name}</td>
                                    )}
                                    <td>
                                        <TooltipActionButton
                                            variantKey="open"
                                            buttonClass={invalid ? "text-danger" : ""}
                                            onClick={() => setSelectedGenerationId(item.id)} />
                                    </td>
                                </tr>
                            );
                        }}
                    </ReorderableTableBody>
                </Table>
            </Row>
            {
                selectedGenerationId &&
                <UniformgenerationOffcanvas
                    uniformTypeId={uniformType.id}
                    usingSizes={uniformType.usingSizes}
                    onHide={() => setSelectedGenerationId(null)}
                    generation={uniformType.uniformGenerationList?.find(g => g.id === selectedGenerationId) ?? null}
                />
            }
        </div >
    )
}