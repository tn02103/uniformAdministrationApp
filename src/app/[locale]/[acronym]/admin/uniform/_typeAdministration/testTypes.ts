import { UniformType } from "@/types/globalUniformTypes";

const sizeListIds = [
    'e667d674-7df8-436b-a2b8-77b06e063d36',
    'a961545b-28a7-409e-9200-1d85ccd53522',
    '07de1d59-4fc6-447b-98a6-da916e5792ef',
];
export const testTypes: UniformType[] = [
    {
        id: "fbe495ec-799e-46c9-8aa5-ca34e1447bf4",
        name: "Test Type",
        acronym: "TT",
        sortOrder: 1,
        usingSizes: true,
        usingGenerations: true,
        issuedDefault: 2,
        fk_defaultSizelist: sizeListIds[0],
        defaultSizelist: {
            id: sizeListIds[0],
            name: "Test Size List",
        },
        uniformGenerationList: [
            {
                id: "ab5d155c-49cf-4019-831e-ca7b3e0bd51c",
                name: "Test Generation 1",
                outdated: false,
                sortOrder: 1,
                fk_sizelist: sizeListIds[0],
                sizelist: {
                    id: sizeListIds[0],
                    name: "Test Size List 1",
                },
            },
            {
                id: "7feb435c-ee23-486d-af9f-b7c874383e22",
                name: "Test Generation 2",
                outdated: false,
                sortOrder: 2,
                fk_sizelist: sizeListIds[1],
                sizelist: {
                    id: sizeListIds[1],
                    name: "Test Size List 2",
                },
            },
            {
                id: "c46adb18-1a04-4dda-884d-972a7aa0e0da",
                name: "Test Generation 3",
                outdated: true,
                sortOrder: 3,
                fk_sizelist: sizeListIds[1],
                sizelist: {
                    id: sizeListIds[1],
                    name: "Test Size List 2",
                },
            },
        ],
    }, {
        id: "c0f5f7d4-5e8b-4c1b-9f2a-7d7e7c6e3f1d",
        name: "Test Type 2",
        acronym: "TT2",
        sortOrder: 2,
        usingSizes: false,
        usingGenerations: false,
        issuedDefault: 1,
        fk_defaultSizelist: sizeListIds[2],
        defaultSizelist: {
            id: sizeListIds[2],
            name: "Test Size List 3",
        },
        uniformGenerationList: [],
    }, {
        id: "26c469f7-aae4-4398-8aba-5d131178c6f8",
        name: "Test Type 3",
        acronym: "TT3",
        sortOrder: 3,
        usingSizes: true,
        usingGenerations: false,
        issuedDefault: 0,
        fk_defaultSizelist: sizeListIds[0],
        defaultSizelist: {
            id: sizeListIds[0],
            name: "Test Size List 1",
        },
        uniformGenerationList: [
            {
                id: "ab5d155c-49cf-4019-831e-ca7b3e0bd51c",
                name: "Test Generation 1",
                outdated: false,
                sortOrder: 1,
                fk_sizelist: sizeListIds[0],
                sizelist: {
                    id: sizeListIds[0],
                    name: "Test Size List 1",
                },
            }
        ],
    }, {
        id: "d1e8b3c2-5a2b-4f8b-9c6e-7d7e7c6e3f1d",
        name: "Test Type 4",
        acronym: "TT4",
        sortOrder: 4,
        usingSizes: true,
        usingGenerations: true,
        issuedDefault: 0,
        fk_defaultSizelist: sizeListIds[0],
        defaultSizelist: {
            id: sizeListIds[0],
            name: "Test Size List 1",
        },
        uniformGenerationList: [
            {
                id: "ab5d155c-49cf-4019-831e-ca7b3e0bd51c",
                name: "Test Generation 1",
                outdated: false,
                sortOrder: 1,
                fk_sizelist: sizeListIds[0],
                sizelist: {
                    id: sizeListIds[0],
                    name: "Test Size List 1",
                },
            },
            {
                id: "7feb435c-ee23-486d-af9f-b7c874383e22",
                name: "Test Generation 2",
                outdated: false,
                sortOrder: 2,
                fk_sizelist: sizeListIds[1],
                sizelist: {
                    id: sizeListIds[1],
                    name: "Test Size List 2",
                },
            }
        ],
    }
]