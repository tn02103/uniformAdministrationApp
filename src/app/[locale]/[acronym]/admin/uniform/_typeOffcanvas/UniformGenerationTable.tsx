import { ReorderableTableBody } from "@/components/reorderDnD/ReorderableTableBody";
import { changeUniformGenerationSortOrder } from "@/dal/uniform/generation/_index";
import { useUniformTypeList } from "@/dataFetcher/uniformAdmin";
import { UniformType, UniformGeneration } from "@/types/globalUniformTypes";
import { faBars } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { useState } from "react";
import { Row, Table, Dropdown } from "react-bootstrap";
import { UniformgenerationOffcanvas } from "./UniformGenerationOffcanvas";
import { TooltipActionButton } from "@/components/Buttons/TooltipIconButton";

type UniformGenerationTableProps = {
    uniformType: UniformType;
}

export const UniformGenerationTable = ({ uniformType }: UniformGenerationTableProps) => {
    console.log("ðŸš€ ~ UniformGenerationTable ~ uniformType:", uniformType)
    const { mutate } = useUniformTypeList();
    const [selectedGenerationId, setSelectedGenerationId] = useState<string | null>(null);

    const handleChangeSortorder = (newArray: UniformGeneration[], itemId: string) => {
        const newPosition = newArray.findIndex(i => i.id === itemId);
        console.log(newArray, itemId, newPosition);
        mutate(changeUniformGenerationSortOrder({ id: itemId, newPosition }));
    };

    return (
        <div>
            <h3 className="text-center">Generationen</h3>
            <hr className="my-0" />
            <Row>
                <Table>
                    <thead>
                        <tr>
                            <th></th>
                            <th>Generation</th>
                            <th className={uniformType.usingSizes ? "d-none d-sm-table-cell" : ""}>Veraltet</th>
                            {uniformType.usingSizes && (
                                <th>GrÃ¶ÃŸenliste</th>
                            )}
                            <th><TooltipActionButton variantKey="create" onClick={() => setSelectedGenerationId('new')} /></th>
                        </tr>
                    </thead>
                    <ReorderableTableBody<UniformGeneration> items={uniformType.uniformGenerationList ?? []} itemType="UNIFORM_GENERATION" onDragEnd={handleChangeSortorder}>
                        {({ draggableRef, previewRef, isDragging, item }) => {
                            const invalid = (uniformType.usingSizes && !item.fk_sizelist)
                            return (
                                <tr
                                    key={item.id}
                                    ref={previewRef}
                                    style={isDragging ? { opacity: 0 } : undefined}
                                >
                                    <td className={invalid ? "text-danger" : ""}>
                                        <span ref={draggableRef} className="p-2">
                                            <FontAwesomeIcon
                                                icon={faBars}
                                                className="text-seccondary"
                                            />
                                        </span>
                                    </td>
                                    <td className={invalid ? "text-danger" : ""}>{item.name}</td>
                                    <td className={invalid ? "text-danger" : "" + uniformType.usingSizes ? "d-none d-sm-table-cell" : ""}>
                                        {item.outdated ? "Ja" : "Nein"}
                                    </td>
                                    {uniformType.usingSizes && (
                                        <td>{item.sizelist?.name}</td>
                                    )}
                                    <td>
                                        <TooltipActionButton
                                            variantKey="open"
                                            buttonClass={invalid ? "text-danger" : ""}
                                            onClick={() => setSelectedGenerationId(item.id)} />
                                    </td>
                                </tr>
                            );
                        }}
                    </ReorderableTableBody>
                </Table>
            </Row>
            {
                selectedGenerationId &&
                <UniformgenerationOffcanvas
                    uniformTypeId={uniformType.id}
                    usingSizes={uniformType.usingSizes}
                    onHide={() => setSelectedGenerationId(null)}
                    generation={uniformType.uniformGenerationList?.find(g => g.id === selectedGenerationId) ?? null}
                />
            }
        </div >
    )
}