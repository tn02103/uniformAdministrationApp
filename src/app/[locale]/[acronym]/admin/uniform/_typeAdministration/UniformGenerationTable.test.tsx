import { UniformType, UniformGeneration } from "@/types/globalUniformTypes";
import { render, screen } from "@testing-library/react";
import { UniformGenerationTable } from "./UniformGenerationTable";
import userEvent from "@testing-library/user-event";
import { testTypes } from "./testTypes";


const testType = testTypes[0];

jest.mock("@/dataFetcher/uniformAdmin", () => {
    const typeListMutate = jest.fn(async (a) => {return a; });
    return {
        useUniformTypeList: jest.fn(() => ({
            mutate: typeListMutate,
        })),
    };
});
jest.mock("@/dal/uniform/generation/_index", () => {
    return {
        changeUniformGenerationSortOrder: jest.fn(() => "uniform generation sortOrder changed"),
    };
});
jest.mock("./UniformGenerationOffcanvas", () => {
    const mock = jest.fn(({ onHide }) => <div data-testid="generationOffcanvasMock" onClick={onHide}>Generation Offcanvas</div>);
    return {
        UniformgenerationOffcanvas: mock,
    };
});
let onDragEndFunction: undefined | ((newArray: UniformGeneration[], itemId: string) => Promise<any>) = undefined;
jest.mock("@/components/reorderDnD/ReorderableTableBody", () => {
    return {
        ReorderableTableBody: jest.fn(({ items, itemType, onDragEnd, children }) => {
            onDragEndFunction = onDragEnd;
            return (
                <tbody data-testid="reorderable-table-body">
                    {items.map(((item: any) => children({ item, draggableRef: undefined, previewRef: undefined, isDragging: false })))}
                </tbody>
            )
        }),
    };
});

describe('<UniformGenerationTable />', () => {
    const { UniformgenerationOffcanvas } = require("./UniformGenerationOffcanvas");
    const { changeUniformGenerationSortOrder } = require("@/dal/uniform/generation/_index");
    const { useUniformTypeList } = require("@/dataFetcher/uniformAdmin");

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render the UniformGenerationTable component', () => {
        render(<UniformGenerationTable uniformType={testType} />);

        expect(screen.getByTestId("uniform-generation-table")).toBeInTheDocument();
        expect(screen.getByTestId("uniform-generation-table")).toMatchSnapshot();
    });

    it('should mark generation as invalid if fk_sizelist is null and usingSizes', () => {
        const testTypeWithInvalidGeneration: UniformType = {
            ...testType,
            uniformGenerationList: [
                ...testType.uniformGenerationList,
                {
                    id: "invalid-generation",
                    name: "Invalid Generation",
                    outdated: false,
                    sortOrder: 4,
                    fk_sizelist: null,
                    sizelist: null,
                },
            ],
        };
        const { unmount } = render(<UniformGenerationTable uniformType={testTypeWithInvalidGeneration} />);

        const invalidGenerationRow = screen.getByText("Invalid Generation").closest("tr");
        expect(invalidGenerationRow?.childNodes[0]).toHaveClass("text-danger");
        expect(invalidGenerationRow?.childNodes[1]).toHaveClass("text-danger");
        expect(invalidGenerationRow?.childNodes[2]).toHaveClass("text-danger");
        expect(invalidGenerationRow?.getElementsByTagName("button")[0]).toHaveClass("text-danger");

        unmount();
        testTypeWithInvalidGeneration.usingSizes = false;
        render(<UniformGenerationTable uniformType={testTypeWithInvalidGeneration} />);

        const validGenerationRow = screen.getByText("Invalid Generation").closest("tr");
        expect(validGenerationRow?.childNodes[0]).not.toHaveClass("text-danger");
        expect(validGenerationRow?.childNodes[1]).not.toHaveClass("text-danger");
        expect(validGenerationRow?.childNodes[2]).not.toHaveClass("text-danger");
        expect(validGenerationRow?.getElementsByTagName("button")[0]).not.toHaveClass("text-danger");
    });

    it('should open the generation offcanvas when clicking the open button', async () => {
        const user = userEvent.setup();
        render(<UniformGenerationTable uniformType={testType} />);

        const openButton = screen.getByText("Test Generation 1").closest("tr")?.getElementsByTagName("button")[0];
        expect(openButton).toBeDefined();
        expect(openButton).not.toBeNull();
        await user.click(openButton!);

        expect(screen.getByTestId("generationOffcanvasMock")).toBeInTheDocument();

        expect(UniformgenerationOffcanvas).toHaveBeenCalledTimes(1);
        expect(UniformgenerationOffcanvas).toHaveBeenCalledWith({
            uniformTypeId: testType.id,
            usingSizes: testType.usingSizes,
            onHide: expect.any(Function),
            generation: testType.uniformGenerationList[0],
        }, {});

        expect(screen.getByTestId("generationOffcanvasMock")).toHaveTextContent("Generation Offcanvas");
        await user.click(screen.getByTestId("generationOffcanvasMock"));

        expect(screen.queryByTestId("generationOffcanvasMock")).not.toBeInTheDocument();
    });

    it('should open the generation offcanvas when clicking the create button', async () => {
        const user = userEvent.setup();
        render(<UniformGenerationTable uniformType={testType} />);

        const createButton = screen.getByTestId("btn_create");
        expect(createButton).toBeDefined();
        expect(createButton).not.toBeNull();
        await user.click(createButton!);

        expect(screen.getByTestId("generationOffcanvasMock")).toBeInTheDocument();

        expect(UniformgenerationOffcanvas).toHaveBeenCalledTimes(1);
        expect(UniformgenerationOffcanvas).toHaveBeenCalledWith({
            uniformTypeId: testType.id,
            usingSizes: testType.usingSizes,
            onHide: expect.any(Function),
            generation: null,
        }, {});

        expect(screen.getByTestId("generationOffcanvasMock")).toHaveTextContent("Generation Offcanvas");
        await user.click(screen.getByTestId("generationOffcanvasMock"));

        expect(screen.queryByTestId("generationOffcanvasMock")).not.toBeInTheDocument();
    });

    it('should call sortOrder function when onDragEnd is triggered', async () => {
        const { toast } = require("react-toastify");
        render(<UniformGenerationTable uniformType={testType} />);

        expect(onDragEndFunction).toBeDefined();
        await onDragEndFunction!([
            testType.uniformGenerationList[1],
            testType.uniformGenerationList[0],
            testType.uniformGenerationList[2],
        ], testType.uniformGenerationList[0].id);

        expect(changeUniformGenerationSortOrder).toHaveBeenCalledTimes(1);
        expect(changeUniformGenerationSortOrder).toHaveBeenCalledWith({
            id: testType.uniformGenerationList[0].id,
            newPosition: 1
        });
        expect(useUniformTypeList().mutate).toHaveBeenCalledTimes(1);
        expect(useUniformTypeList().mutate).toHaveBeenCalledWith("uniform generation sortOrder changed");

        expect(toast.success).toHaveBeenCalledTimes(1);
        expect(toast.success).toHaveBeenCalledWith("common.success.changeSortorder");
    });

    it('should catch error when sortOrder function fails', async () => {
        const { toast } = require("react-toastify");
        changeUniformGenerationSortOrder.mockImplementationOnce(async () => { throw new Error("Error") });

        render(<UniformGenerationTable uniformType={testType} />);

        expect(onDragEndFunction).toBeDefined();
        await onDragEndFunction!([
            testType.uniformGenerationList[1],
            testType.uniformGenerationList[0],
            testType.uniformGenerationList[2],
        ], testType.uniformGenerationList[0].id);

        expect(changeUniformGenerationSortOrder).toHaveBeenCalledTimes(1);
        expect(changeUniformGenerationSortOrder).toHaveBeenCalledWith({
            id: testType.uniformGenerationList[0].id,
            newPosition: 1
        });
        expect(useUniformTypeList().mutate).toHaveBeenCalledTimes(1);
        expect(useUniformTypeList().mutate).not.toHaveBeenCalledWith("uniform generation sortOrder changed");

        expect(toast.success).toHaveBeenCalledTimes(0);
        expect(toast.error).toHaveBeenCalledTimes(1);
        expect(toast.error).toHaveBeenCalledWith("common.error.actions.changeSortorder");
    });

    it('should not call sortOrder function when itemId is not in the list', () => {
        render(<UniformGenerationTable uniformType={testType} />);

        expect(onDragEndFunction).toBeDefined();
        onDragEndFunction!([
            testType.uniformGenerationList[1],
            testType.uniformGenerationList[0],
            testType.uniformGenerationList[2],
        ], "invalid-item-id");

        expect(changeUniformGenerationSortOrder).not.toHaveBeenCalled();
    });

    it('should not call sortOrder function when list has wrong size', () => {
        render(<UniformGenerationTable uniformType={testType} />);

        expect(onDragEndFunction).toBeDefined();
        onDragEndFunction!([
            testType.uniformGenerationList[1],
            testType.uniformGenerationList[0],
        ], testType.uniformGenerationList[0].id);

        expect(changeUniformGenerationSortOrder).not.toHaveBeenCalled();
    });

    it('should not show sizeList column when usingSizes is false', () => {
        const testTypeWithoutSizeList: UniformType = {
            ...testType,
            usingSizes: false,
        };
        render(<UniformGenerationTable uniformType={testTypeWithoutSizeList} />);

        expect(screen.queryByText("common.uniform.sizelist.label")).not.toBeInTheDocument();
        const row = screen.getByText("Test Generation 1").closest("tr");
        expect(row).not.toBeNull();
        expect(row!.childNodes).toHaveLength(4); // 3 columns + 1 action column
    });
    it('should show outdated column on Mobile when usingSizes is false ', () => {
        const testTypeWithoutSizeList: UniformType = {
            ...testType,
            usingSizes: false,
        };
        render(<UniformGenerationTable uniformType={testTypeWithoutSizeList} />);

        expect(screen.queryByText("common.uniform.generation.outdated")).toBeInTheDocument();
        expect(screen.queryByText("common.uniform.generation.outdated")).not.toHaveClass("d-none d-sm-table-cell");
        const row = screen.getByText("Test Generation 1").closest("tr");
        expect(row).not.toBeNull();
        expect(row?.childNodes[2]).not.toHaveClass("d-none d-sm-table-cell");
    });
    it('should hide outdated column on Mobile when usingSizes is true ', () => {
        render(<UniformGenerationTable uniformType={testType} />);

        expect(screen.queryByText("common.uniform.generation.outdated")).toBeInTheDocument();
        expect(screen.queryByText("common.uniform.generation.outdated")).toHaveClass("d-none d-sm-table-cell");
        const row = screen.getByText("Test Generation 1").closest("tr");
        expect(row).not.toBeNull();
        expect(row?.childNodes[2]).toHaveClass("d-none d-sm-table-cell");
    });
});