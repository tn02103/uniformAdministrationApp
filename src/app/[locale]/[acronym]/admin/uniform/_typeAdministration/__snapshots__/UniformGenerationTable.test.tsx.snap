// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<UniformGenerationTable /> renders the component 1`] = `
<div
  data-testid="uniform-generation-table"
>
  <h3
    class="text-center"
  >
    common.uniform.generation.label
  </h3>
  <hr
    class="my-0"
  />
  <div
    class="row"
  >
    <table
      class="table"
    >
      <thead>
        <tr>
          <th />
          <th>
            common.uniform.generation.label
          </th>
          <th
            class="d-none d-sm-table-cell"
          >
            common.uniform.generation.outdated
          </th>
          <th>
            common.uniform.sizelist.label
          </th>
          <th>
            <button
              aria-label="create"
              class="border-0 align-self-center undefined btn btn-outline-success btn-sm"
              data-testid="btn_create"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-plus "
                data-icon="plus"
                data-prefix="fas"
                role="img"
                viewBox="0 0 448 512"
              >
                <path
                  d="M256 64c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 160-160 0c-17.7 0-32 14.3-32 32s14.3 32 32 32l160 0 0 160c0 17.7 14.3 32 32 32s32-14.3 32-32l0-160 160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-160 0 0-160z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </th>
        </tr>
      </thead>
      <tbody
        data-testid="reorderable-table-body"
      >
        <tr
          aria-label="Test Generation 1"
        >
          <td
            class=""
          >
            <span
              aria-label="drag item"
              class="p-2 drag-handler"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-bars "
                data-icon="bars"
                data-prefix="fas"
                role="img"
                viewBox="0 0 448 512"
              >
                <path
                  d="M0 96C0 78.3 14.3 64 32 64l384 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 128C14.3 128 0 113.7 0 96zM0 256c0-17.7 14.3-32 32-32l384 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 288c-17.7 0-32-14.3-32-32zM448 416c0 17.7-14.3 32-32 32L32 448c-17.7 0-32-14.3-32-32s14.3-32 32-32l384 0c17.7 0 32 14.3 32 32z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </td>
          <td
            class=""
          >
            Test Generation 1
          </td>
          <td
            class=" d-none d-sm-table-cell"
          >
            common.no
          </td>
          <td>
            Test Size List 1
          </td>
          <td>
            <button
              aria-label="open"
              class="border-0 align-self-center  btn btn-outline-secondary btn-sm"
              data-testid="btn_open"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-arrow-up-right-from-square "
                data-icon="arrow-up-right-from-square"
                data-prefix="fas"
                role="img"
                viewBox="0 0 512 512"
              >
                <path
                  d="M320 0c-17.7 0-32 14.3-32 32s14.3 32 32 32l82.7 0-201.4 201.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L448 109.3 448 192c0 17.7 14.3 32 32 32s32-14.3 32-32l0-160c0-17.7-14.3-32-32-32L320 0zM80 96C35.8 96 0 131.8 0 176L0 432c0 44.2 35.8 80 80 80l256 0c44.2 0 80-35.8 80-80l0-80c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 80c0 8.8-7.2 16-16 16L80 448c-8.8 0-16-7.2-16-16l0-256c0-8.8 7.2-16 16-16l80 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L80 96z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </td>
        </tr>
        <tr
          aria-label="Test Generation 2"
        >
          <td
            class=""
          >
            <span
              aria-label="drag item"
              class="p-2 drag-handler"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-bars "
                data-icon="bars"
                data-prefix="fas"
                role="img"
                viewBox="0 0 448 512"
              >
                <path
                  d="M0 96C0 78.3 14.3 64 32 64l384 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 128C14.3 128 0 113.7 0 96zM0 256c0-17.7 14.3-32 32-32l384 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 288c-17.7 0-32-14.3-32-32zM448 416c0 17.7-14.3 32-32 32L32 448c-17.7 0-32-14.3-32-32s14.3-32 32-32l384 0c17.7 0 32 14.3 32 32z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </td>
          <td
            class=""
          >
            Test Generation 2
          </td>
          <td
            class=" d-none d-sm-table-cell"
          >
            common.no
          </td>
          <td>
            Test Size List 2
          </td>
          <td>
            <button
              aria-label="open"
              class="border-0 align-self-center  btn btn-outline-secondary btn-sm"
              data-testid="btn_open"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-arrow-up-right-from-square "
                data-icon="arrow-up-right-from-square"
                data-prefix="fas"
                role="img"
                viewBox="0 0 512 512"
              >
                <path
                  d="M320 0c-17.7 0-32 14.3-32 32s14.3 32 32 32l82.7 0-201.4 201.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L448 109.3 448 192c0 17.7 14.3 32 32 32s32-14.3 32-32l0-160c0-17.7-14.3-32-32-32L320 0zM80 96C35.8 96 0 131.8 0 176L0 432c0 44.2 35.8 80 80 80l256 0c44.2 0 80-35.8 80-80l0-80c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 80c0 8.8-7.2 16-16 16L80 448c-8.8 0-16-7.2-16-16l0-256c0-8.8 7.2-16 16-16l80 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L80 96z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </td>
        </tr>
        <tr
          aria-label="Test Generation 3"
        >
          <td
            class=""
          >
            <span
              aria-label="drag item"
              class="p-2 drag-handler"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-bars "
                data-icon="bars"
                data-prefix="fas"
                role="img"
                viewBox="0 0 448 512"
              >
                <path
                  d="M0 96C0 78.3 14.3 64 32 64l384 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 128C14.3 128 0 113.7 0 96zM0 256c0-17.7 14.3-32 32-32l384 0c17.7 0 32 14.3 32 32s-14.3 32-32 32L32 288c-17.7 0-32-14.3-32-32zM448 416c0 17.7-14.3 32-32 32L32 448c-17.7 0-32-14.3-32-32s14.3-32 32-32l384 0c17.7 0 32 14.3 32 32z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </td>
          <td
            class=""
          >
            Test Generation 3
          </td>
          <td
            class=" d-none d-sm-table-cell"
          >
            common.yes
          </td>
          <td>
            Test Size List 2
          </td>
          <td>
            <button
              aria-label="open"
              class="border-0 align-self-center  btn btn-outline-secondary btn-sm"
              data-testid="btn_open"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-arrow-up-right-from-square "
                data-icon="arrow-up-right-from-square"
                data-prefix="fas"
                role="img"
                viewBox="0 0 512 512"
              >
                <path
                  d="M320 0c-17.7 0-32 14.3-32 32s14.3 32 32 32l82.7 0-201.4 201.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L448 109.3 448 192c0 17.7 14.3 32 32 32s32-14.3 32-32l0-160c0-17.7-14.3-32-32-32L320 0zM80 96C35.8 96 0 131.8 0 176L0 432c0 44.2 35.8 80 80 80l256 0c44.2 0 80-35.8 80-80l0-80c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 80c0 8.8-7.2 16-16 16L80 448c-8.8 0-16-7.2-16-16l0-256c0-8.8 7.2-16 16-16l80 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L80 96z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
